src/components/logging.ts

```diff
@@ -12,221 +12,372 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+function isMessageNoise(messageOrObjects: MsgType[]) {
+	return !!messageOrObjects.find(messageOrObject => {
+		if (typeof messageOrObject !== "object") {
+			return false;
+		}
+
+		const possibleError = messageOrObject as {
+            error?: string, body?: { error?: string, errcode?: string}, errcode?: string
+        }
+
+		const error = possibleError?.error || possibleError?.body?.error;
+		const errcode = possibleError?.errcode || possibleError?.body?.errcode;
+
+		if (errcode === "M_NOT_FOUND" && error === "Room account data not found") {
+			return true;
+		}
+
+		if (errcode === "M_NOT_FOUND" && error === "Event not found.") {
+			return true;
+		}
+
+		if (errcode === "M_USER_IN_USE") {
+			return true;
+		}
+
+		return false;
+	});
 }
 
-type MessagePart = unknown;
-interface LoggerConfig {
-    console?: LogLevel|"off",
-    fileDatePattern?: string,
-    timestampFormat?: string,
-    files?: {
-        [filename: string]: LogLevel|"off",
-    }
-    maxFiles?: number,
+interface LogEntry extends winston.Logform.TransformableInfo {
+    data: MsgType[];
+    requestId: string;
+    module: string;
 }
 
-export class LogWrapper {
-    private logger: Logger|null = null;
-    private messages: {type: LogLevel, message: string}[] = [];
 
-    public setLogger(logger: Logger): void {
-        this.logger = logger;
-    }
+export interface CustomLogger {
+    verbose: (message: string, ...metadata: any[]) => void,
+    debug: (message: string, ...metadata: any[]) => void,
+    info: (message: string, ...metadata: any[]) => void,
+    warn: (message: string, ...metadata: any[]) => void,
+    error: (message: string, ...metadata: any[]) => void,
+}
 
-    public debug(...messageParts: MessagePart[]): void { this.log(messageParts, 'debug') }
+export type LogLevel = "debug"|"info"|"warn"|"error"|"trace";
 
-    public info(...messageParts: MessagePart[]): void { this.log(messageParts, 'info') }
+export interface LoggingOpts {
+    /**
+     * The log level used by the console output.
+     */
+    console?: "debug"|"info"|"warn"|"error"|"trace"|"off";
+    /**
+     * Should the logs be outputted in JSON format, for consumption by a collector.
+     */
+    json?: boolean;
+    /**
+     * Should the logs color-code the level strings in the output.
+     */
+    colorize?: boolean;
+    /**
+     * Timestamp format used in the log output.
+     * @default "HH:mm:ss:SSS"
+     */
+    timestampFormat?: string;
+}
+
+export interface LoggingOptsFile extends LoggingOpts {
+    /**
+     * An object mapping a file name to a logging level. The file will contain
+     * all logs for that level inclusive up to the highest level. (`info` will contain `warn`, `error` etc)
+     * Use `%DATE%` to set the date of the file within the string.
+     * Use the `fileDatePattern` to set the date format.
+     * @example {"info-%DATE%.log": "info"}
+     */
+    files: {
+        [filename: string]: LogLevel,
```

Comment:
/me starts naming his log files `__proto__`. :foot::gun:

Reply:
AH! Yes!

Reply:
This is difficult, it's an interface to the logger config so we can't make changes lightly. This is also only really configured by administrators and ultimately comes from a yaml file somewhere. I feel like we're relatively safe from `__proto__` corruption here:

1. You'd have to as an admin explicitly put this key in, which feels like something you'd never really do.
2. You'd only be harming your own software in doing so.

It does feel a bit strange to ask people to submit a map as a parameter to a function if it needs to be parsed from an object first during YAML parsing, so I'm hesistant to say we should change this.